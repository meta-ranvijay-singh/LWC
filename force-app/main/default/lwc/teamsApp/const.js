const SEVERITY = {
  success: "success",
  error: "error"
};

const TOAST_MODE = {
  dismissable: "dismissable"
};

export { SEVERITY, TOAST_MODE };
