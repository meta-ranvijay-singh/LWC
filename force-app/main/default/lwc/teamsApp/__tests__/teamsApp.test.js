import TeamsApp from "c/teamsApp";
//import { mockToast } from "@ffdev/lwc-test-utils/lightning/platformShowToast";
import { ShowToastEventName } from "lightning/platformShowToastEvent";
import { getCreateComponent, flushPromises, clearUp } from "c/test_utils";
import { SEVERITY, TOAST_MODE } from "../const";
import getTeamList from "@salesforce/apex/CustomController.getTeamList";
import getAllTeamMembers from "@salesforce/apex/CustomController.getAllTeamMembers";
import insertMember from "@salesforce/apex/CustomController.insertMember";

const TeamsMockData = require("./data/teamsData.json");
const TeamsMemeberMockData = require("./data/teamMemberData.json");

const createComponent = getCreateComponent(TeamsApp, "c-teams-app");

//jest.mock("lightning/platformShowToastEvent", () => mockToast);

jest.mock(
  "@salesforce/apex/CustomController.getTeamList",
  () => {
    const { createApexTestWireAdapter } = require("@salesforce/sfdx-lwc-jest");
    return {
      default: createApexTestWireAdapter(jest.fn())
    };
  },
  { virtual: true }
);

jest.mock(
  "@salesforce/apex/CustomController.getAllTeamMembers",
  () => {
    const { createApexTestWireAdapter } = require("@salesforce/sfdx-lwc-jest");
    return {
      default: createApexTestWireAdapter(jest.fn())
    };
  },
  { virtual: true }
);

jest.mock(
  "@salesforce/apex/CustomController.insertMember",
  () => {
    return {
      default: jest.fn()
    };
  },
  { virtual: true }
);

describe("c--teams-app", () => {
  afterEach(clearUp);

  describe("Create component", () => {
    it("renders component correctly", () => {
      // When
      const { component } = createComponent();

      getTeamList.emit(TeamsMockData);
      getAllTeamMembers.emit(TeamsMemeberMockData);

      // Then
      expect(component).toMatchSnapshot();
    });
  });

  describe("Handle events", () => {
    it("verify data insert successfully", async () => {
      // When
      const { component, root } = createComponent();
      let mockMember = {
        Id: "1234",
        Name: "Harry",
        Team__c: TeamsMockData[0].Id,
        Team__r: {
          Name: TeamsMockData[0].Name
        },
        Skills__c: "Batsman"
      };

      getTeamList.emit(TeamsMockData);
      getAllTeamMembers.emit(TeamsMemeberMockData);
      insertMember.mockResolvedValue(mockMember);

      const showToastEventHandler = jest.fn();
      component.addEventListener(ShowToastEventName, showToastEventHandler);

      let memberSkillComponent = root.querySelector("c-member-skills");
      memberSkillComponent.dispatchEvent(
        new CustomEvent("insert", {
          detail: {
            memberName: mockMember.Name,
            teamId: mockMember.Team__c + ":" + mockMember.Team__r.Name,
            memberSkills: mockMember.Skills__c
          }
        })
      );

      // Then
      await flushPromises();
      expect(insertMember).toBeCalledTimes(1);
      expect(showToastEventHandler).toHaveBeenCalled();
      expect(showToastEventHandler.mock.calls[0][0].detail.variant).toBe(
        SEVERITY.success
      );
      expect(showToastEventHandler.mock.calls[0][0].detail.mode).toBe(
        TOAST_MODE.dismissable
      );
    });

    it("verify data insert failure", async () => {
      // When
      const { component, root } = createComponent();

      getTeamList.emit(TeamsMockData);
      getAllTeamMembers.emit(TeamsMemeberMockData);
      insertMember.mockResolvedValue(false);

      const showToastEventHandler = jest.fn();
      component.addEventListener(ShowToastEventName, showToastEventHandler);

      let memberSkillComponent = root.querySelector("c-member-skills");
      memberSkillComponent.dispatchEvent(
        new CustomEvent("insert", {
          detail: {
            memberName: "Harry",
            teamId: TeamsMockData[0].Id + ":" + TeamsMockData[0].Name,
            memberSkills: "Batsman"
          }
        })
      );

      // Then
      await flushPromises();
      expect(insertMember).toBeCalledTimes(1);
      expect(showToastEventHandler).toHaveBeenCalled();
      expect(showToastEventHandler.mock.calls[0][0].detail.variant).toBe(
        SEVERITY.error
      );
      expect(showToastEventHandler.mock.calls[0][0].detail.mode).toBe(
        TOAST_MODE.dismissable
      );
    });
  });
});
