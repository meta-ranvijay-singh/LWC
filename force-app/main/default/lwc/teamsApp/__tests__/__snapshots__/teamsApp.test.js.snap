// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c--teams-app Create component renders component correctly 1`] = `
<c-teams-app>
  #shadow-root(open)
    <lightning-card
      class="slds-p-around_medium"
    >
      #shadow-root(open)
        <slot />
        <slot
          name="actions"
        />
        <slot
          name="footer"
        />
        <slot
          name="title"
        />
      <c-member-skills>
        #shadow-root(open)
          <lightning-card
            class="slds-size_medium slds-p-around_medium"
          >
            #shadow-root(open)
              <slot />
              <slot
                name="actions"
              />
              <slot
                name="footer"
              />
              <slot
                name="title"
              />
            <lightning-input
              class="slds-size_1-of-3 slds-align_absolute-center"
            >
              #shadow-root(open)
            </lightning-input>
            <lightning-combobox
              class="slds-size_1-of-3 slds-align_absolute-center"
            >
              #shadow-root(open)
            </lightning-combobox>
            <lightning-input
              class="slds-size_1-of-3 slds-align_absolute-center"
            >
              #shadow-root(open)
            </lightning-input>
            <div
              slot="footer"
            >
              <lightning-button
                class="slds-m-left_x-small"
                title="Primary action"
              >
                #shadow-root(open)
              </lightning-button>
            </div>
          </lightning-card>
      </c-member-skills>
      <c-teams-list>
        #shadow-root(open)
          <lightning-card
            class="slds-size_medium slds-p-around_medium"
          >
            #shadow-root(open)
              <slot />
              <slot
                name="actions"
              />
              <slot
                name="footer"
              />
              <slot
                name="title"
              />
            <lightning-combobox
              class="slds-size_1-of-3 slds-align_absolute-center"
            >
              #shadow-root(open)
            </lightning-combobox>
          </lightning-card>
      </c-teams-list>
    </lightning-card>
</c-teams-app>
`;
